### PR TITLE
Fix split MongoDB Panache packages

### PR DESCRIPTION
**Pull request by loicmathieu**
Related to #18732

Will fix these warnings:

```
    "io.quarkus.mongodb.panache" found in [io.quarkus:quarkus-mongodb-panache-common, io.quarkus:quarkus-mongodb-panache]
    "io.quarkus.mongodb.panache.reactive" found in [io.quarkus:quarkus-mongodb-panache-common, io.quarkus:quarkus-mongodb-panache]
    "io.quarkus.mongodb.panache.reactive.runtime" found in [io.quarkus:quarkus-mongodb-panache, io.quarkus:quarkus-mongodb-panache-common]
    "io.quarkus.mongodb.panache.runtime" found in [io.quarkus:quarkus-mongodb-panache-common, io.quarkus:quarkus-mongodb-panache]
```

### CODE REVIEW

**Comment by loicmathieu** [comment]
@gsmet the current impl remove the split packages on the runtime and deployment packages.

To remove the split packages on the io.quarkus.mongodb.panache and io.quarkus.mongodb.panache.reactive packages, I need to duplicate the annotations (easy, already done) and to move some interfaces.
This will create a breaking change on a lot of code as I would need to move the PanacheQuery interface that almost all user of the extension use ! Is it worth it ?

File: extensions/panache/mongodb-panache-common/runtime/src/test/java/io/quarkus/mongodb/panache/common/runtime/MongoOperationsTest.java [hunk 1]
```diff
@@ -1,6 +1,6 @@
-package io.quarkus.mongodb.panache.runtime;
+package io.quarkus.mongodb.panache.common.runtime;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (gsmet) [inline extensions/panache/mongodb-panache-common/runtime/src/test/java/io/quarkus/mongodb/panache/common/runtime/MongoOperationsTest.java:3]: Please avoid star imports.

**Review by gsmet — COMMENTED** [review]
> This will create a breaking change on a lot of code as I would need to move the PanacheQuery interface that almost all user of the extension use ! Is it worth it ?

Well, that's not a question of "is it worth it?" but more a question of "can we do without it?". Split packages come with all sorts of problem so we really need to avoid them.

I think it's good enough if people have one version to move from one class to the other.

We need to add a paragraph in the migration guide about these changes.

**Comment by loicmathieu** [comment]
> I think it's good enough if people have one version to move from one class to the other.

This is the issue, they will not have one version to move from a class to the other, they will need to do it immediatly as I need to move some interfaces so I cannot duplicate them as I did for the annotations.

**Comment by gsmet** [comment]
Why can't you keep the original interfaces and have the old one extending the new one? That's how we do things when moving interfaces around.

**Comment by loicmathieu** [comment]
I was thinking something similar today. It should work. I'll try this tomorrow

**Comment by loicmathieu** [comment]
@gsmet it's the other way around, the new interfaces need to extends the old one to make old code compile.
Can you have a quick look at the current impl of this PR ? It should works (test still runs).

I still need to:
- Polish the code
- Add the package exclusion to avoid the WARNING
- Make both `@MongoEntity` annotation works
- Add some test
- Check the documentation
- Add a note on the upgrading page

**Comment by loicmathieu** [comment]
@gsmet I finished the PR, this has been a little more complicated than expected because of the usage of the MongoEntity annotation.

Maybe @FroMage or @evanchooly can also have a look as it's a pretty huge PR.

Now there is still a bunch of packages inside mongodb-panache-common that are not child of the io.quarkus.mongodb.panache.common package. There is currently no issue with them so I'll move them inside the common package on a followup PR to have a cleaner package hierarchy.

I'll write something in the upgrade page when the PR will be merged.

**Comment by evanchooly** [comment]
Overall it looks fine to me.  Largely just a package move and concomitant code support changes.  I'm curious why old-extending-new didn't work as that would be how I would've done it, too.  But otherwise looks like a fairly mechanical rearrangement to me.

**Comment by loicmathieu** [comment]
@evanchooly the other way around didn't compile in Kotlin:

```
[ERROR] /data/git/quarkus/extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/Book.kt:[158,15] Type mismatch: inferred type is io.quarkus.mongodb.panache.common.PanacheUpdate! but io.quarkus.mongodb.panache.PanacheUpdate was expected
```

**Review by gsmet — APPROVED** [review]
OK, let's make progress on this.

I will backport it as I want 2.2 to be clean on this side so we need to push it to 2.1 first.

**Comment by gsmet** [comment]
@loicmathieu can you add your classes there: https://github.com/quarkusio/quarkus/wiki/Migration-Guide-2.1#panache-split-packages ? Thanks!

**Comment by loicmathieu** [comment]
@gsmet I'll do it tomorrow. 
But I didn't merged it because one test didn't pass and I don't understand why. Seems to be linked to devservices.
I didn't had the time to check it today so I fear it will fail the CI. 
Feel free to ignore the failing test I'l have a look tomorrow

**Comment by loicmathieu** [comment]
@gsmet I updated the wiki page.
I think the old calsses should be kept more than one release as I don't think everybody upgrade Quarkus version on each release so we should remove them in 3 or 4 releases.

**Pull request merged**